framework/Optimizers/gradients/FiniteDifference.py

```diff
@@ -70,17 +70,98 @@ def chooseEvaluationPoints(self, opt, stepSize):
 
     directions = np.atleast_1d(randomUtils.random(self.N) < 0.5) * 2 - 1
     for o, optVar in enumerate(self._optVars):
+      # pick a new grad eval point
       optValue = opt[optVar]
       new = copy.deepcopy(opt)
-      delta = dh * directions[o]
+      delta = dh * directions[o] # note this is NORMALIZED space delta
       new[optVar] = optValue + delta
+      # constraint handling
+      if constraints is not None:
+        # all constraints speak DENORM space, not NORM space
+        denormed = constraints['denormalize'](new)
+        altPoint = self._handleConstraints(denormed, constraints['denormalize'](opt), optVar, constraints)
+        # need NORM point, delta
+        new = constraints['normalize'](altPoint)
+        delta = new[optVar] - opt[optVar]
+      # store as samplable point
       evalPoints.append(new)
       evalInfo.append({'type': 'grad',
                        'optVar': optVar,
                        'delta': delta})
+    return evalPoints, evalInfo
 
+  def _handleConstraints(self, newPoint, original, optVar, constraints):
+    """
+      Allows the FiniteDifference to handle grad points that might violate constraints.
+      Note this should be generalized as much as possible to the base class, if the different
+      gradient approximation algorithms can find common ground in this algorithm.
```

Comment:
do you think we can try to generalize this in this PR? (maybe @dylanjm  can do it if you do not have time now?)

Reply:
I do think we can; when we start in on the next work for the HERON milestone, this is on the top of the list of things to include in the work.